nd2reader/parser.py

```diff
@@ -283,7 +283,13 @@ def _get_raw_image_data(self, image_group_number, channel_offset, height, width)
         if np.any(image_data):
             return timestamp, Frame(image_data, metadata=self._get_frame_metadata())
 
-        raise NoImageError
+        # If a blank "gap" image is encountered, generate an empty frame (black image) of corresponding height and width to avoid 
+        # errors with ND2-files with missing frames.
+        else: 
+            empty_frame = np.zeros([height, width])
```

Comment:
I think an image of zeros would not be the best way to handle this... IMHO it would be clearer that there is no data if it is an image consisting of `np.nan`.  Could you replace this with `np.full((height, width), np.nan)`? Thanks!

Reply:
Hi Ruben - I see the appeal for using np.nan to reflect the fact that data is missing. However, nan-type is not expected in most image processing pipelines/software and leads to unexpected behavior downstream, the reason which may not be obvious for the average user (see below). E.g. many numpy operations with np.nan yield nan: 

```
>>> n = np.nan
>>> s = [3,27,np.nan,93.3]

>>> print(np.divide(100,n))
>>> print(np.add(100,n))
>>> print(np.power(100,n))
>>> print(np.mean(s))
>>> print(np.max(s))
nan
nan
nan
nan
nan
```
One more concrete example: a common operation in image-processing is to re-scale intensity to fall between 0-1, or to normalize image values to have zero-mean. Both operations require division by image max and mean respectively. If max/mean is calculated across a dimension that contains `nan`, all values within the stack will convert to `nan`, effectively deleting all data. This happens e.g. when calling` skimage.exposure.rescale_intensity()` since image min and max are calculated by numpy functions: https://github.com/scikit-image/scikit-image/blob/d8f23c64c45427962ec5e7f975140de2d726547d/skimage/exposure/exposure.py#L258-L259

The same risk exists for virtually all arithmetic operations called on arrays that include nan-types. 

**In effect, serving nan-type will force most users to manually convert** `nan ` **to zeros before running computations.** I think missing frames are handled better as zeros than nan-type. Most packages for writing image-files generally convert nan-type to zeros as well.

LMK what you think. 

I will clean up the `NoImageError()` and update the PR.

Reply:
Hi Wolfgang,
I much appreciate your thorough reply!

While I certainly understand your concerns, I still prefer to return an image of `np.nan`'s. The reason is that in this way, it is clear to the user that this is an empty 'gap' image. While it is a trade-off, I still think it is best to put the responsibility for dealing with this missing data with the user.

I think using the `np.nan` value is actually a convention for a lot of Numpy/Pandas/Scipy libraries. Consider for example this except from a book that deals with the question on how Pandas handles missing data (https://jakevdp.github.io/PythonDataScienceHandbook/03.04-missing-values.html)

> Generally, they revolve around one of two strategies: using a mask that globally indicates missing values, or choosing a sentinel value that indicates a missing entry. (...)
> In the sentinel approach, the sentinel value could be some data-specific convention, such as indicating a missing integer value with -9999 or some rare bit pattern, or it could be a more global convention, such as indicating a missing floating-point value with NaN (Not a Number), a special value which is part of the IEEE floating-point specification. (...)
> NumPy does have support for masked arrays – that is, arrays that have a separate Boolean mask array attached for marking data as "good" or "bad." Pandas could have derived from this, but the overhead in both storage, computation, and code maintenance makes that an unattractive choice.
> 
> With these constraints in mind, Pandas chose to use sentinels for missing data (...)
> NaN (acronym for Not a Number), is different; it is a special floating-point value recognized by all systems that use the standard IEEE floating-point representation (...)
> Notice that NumPy chose a native floating-point type for this array: this means that unlike the object array from before, this array supports fast operations pushed into compiled code. You should be aware that NaN is a bit like a data virus–it infects any other object it touches. Regardless of the operation, the result of arithmetic with NaN will be another NaN.

So to summarize, although it requires a bit more care on the side of the user, I think it is a good convention to use `np.nan`. There are different ways to handle the `NaN`s further down the line, for example using the aggregate functions `np.nansum, np.nanmean, np.nanmax` etc. or using masked arrays. But I think it is up to the user on how to handle these missing frames. The advantage of `NaN`s is that the user can clearly see something is going wrong if they don't consider the gap frames.

So I still really want to return an array of `NaN`s. If you want, you can add a warning like this:
`warnings.warn('ND2 file contains gap frames which are represented by np.nan-filled arrays')`
In this way, it is really unambiguous to the user what is happening with the data.

Reply:
Hey Ruben,

I think adding a warning is a good compromise! 

I modified the PR to reflect your suggestion, although maybe we could add another hint on how to handle np.nan-filled arrays?

`warnings.warn('ND2 file contains gap frames which are represented by np.nan-filled arrays; to convert to zeros use e.g. np.nan_to_num(array)')`

I also found two other references to NoImageError in parser.py that I missed previously:

https://github.com/rbnvrw/nd2reader/blob/9e0358882c0a6da6e267a50bbeb75e569f403099/nd2reader/parser.py#L75-L76

https://github.com/rbnvrw/nd2reader/blob/9e0358882c0a6da6e267a50bbeb75e569f403099/nd2reader/parser.py#L104-L105

I removed the reference to NoImageError in the exceptions. Should we remove the exceptions altogether? 

Bests, 

Wolfgang

Reply:
Hi Wolfgang,
Good that you spotted the other two instances!
I like your more extensive/explanatory warning, if you add that too then I'll merge the PR.
Thanks again for your efforts! :1st_place_medal:

Reply:
Done! Thanks Ruben 👍